packages/databus/src/index.ts

```diff
@@ -0,0 +1,135 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) Jupyter Development Team.
+| Distributed under the terms of the Modified BSD License.
+|----------------------------------------------------------------------------*/
+
+import { DisposableDelegate, IDisposable } from '@phosphor/disposable';
+
+import { ArrayExt } from '@phosphor/algorithm';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+/**
+ * An interface for an abstract dataset, with a mimetype, data, and metadata.
+ *
+ * This is templated over the type of data `T` and the type of metadata `U`.
+ *
+ * #### Notes
+ * This interface is similar conceptually to that of nbformat.IMimeBundle,
+ * but with only a single mimetype and different structure. We expect there
+ * to be utilities to convert between this formats.
+ */
+export interface IDataset<T, U> {
+  /**
+   * The string mimetype for the dataset.
+   */
+  mimeType: string;
+
+  /***
+   * The data of the dataset.
+   *
+   * #### Notes
+   * The notion of data in the databus is completely abstract, and typed as an
+   * any. Publishers and consumers of a given mimetype will need to agree
+   * what the actual data is and cast it appropriately before use.
+   */
+  data: T;
+
+  /**
+   * The metadata associate with the dataset.
+   *
+   * #### Notes
+   * Like the data itself, the databus notion of metadata is entirely abstract.
+   * We are currently exploring different metadata schemes, so this may change
+   * in the future.
+   */
+  metadata: U;
+}
+
+/**
+ * A databus object, for managing datasets in JupyterLab.
+ */
+export class DataBus {
+  /**
+   * Construct a new databus.
+   */
+  constructor() {}
+
+  /**
+   * Publish a dataset to the databus.
+   *
+   * @param dataset - the `IDataset` to publish to the databus.
+   *
+   * @returns A disposable which will remove the dataset.
+   *
+   * @throws An error if the given dataset is already published.
+   */
+  publish(dataset: IDataset<any, any>): IDisposable {
+    if (this._datasets.find(i => i === dataset) === undefined) {
+      throw new Error(`Dataset already published`);
+    }
+
+    this._datasets.push(dataset);
+
+    this._datasetsChanged.emit({ dataset, type: 'added' });
+
+    return new DisposableDelegate(() => {
```

Comment:
A consequence of this pattern is that only the publisher of a dataset will be able to remove it. I think that makes good sense, just trying to think through the lifecycle a bit...

Reply:
Yes, I thought of that as well. I can imagine situations where others might want to remove a dataset, but I also don't want to over engineer it. Ideas on an alterative?